topdown/builtins/builtins.go

```diff
@@ -28,6 +29,67 @@ func (c Cache) Get(k interface{}) (interface{}, bool) {
 	return v, ok
 }
 
+// We use an ast.Object for the cached keys/values because a naive
+// map[ast.Value]ast.Value will not correctly detect value equality of
+// the member keys.
+type NDBCache map[string]ast.Object
+
+// Put updates the cache for the named built-in.
+// Automatically creates the 2-level hierarchy as needed.
+func (c NDBCache) Put(name string, k, v ast.Value) {
+	if _, ok := c[name]; !ok {
+		c[name] = ast.NewObject()
+	}
+	c[name].Insert(ast.NewTerm(k), ast.NewTerm(v))
+}
+
+// Get returns the cached value for k for the named builtin.
+func (c NDBCache) Get(name string, k ast.Value) (ast.Value, bool) {
+	if m, ok := c[name]; ok {
+		v := m.Get(ast.NewTerm(k))
+		if v != nil {
+			return v.Value, true
+		}
+		return nil, false
+	}
+	return nil, false
+}
+
+// Convenience functions for serializing the data structure.
+func (c NDBCache) MarshalJSON() ([]byte, error) {
+	out := make(map[string]json.RawMessage)
```

Comment:
I didn't expect it either, but if I don't provide a customized `MarshalJSON` implementation, I get a: `json: cannot unmarshal array into Go value of type map[string]interface {}` error during the marshal/unmarshal test in `rego_test.go`.

I suspect this is happening because raw `ast.Value` types do not directly implement `Marshal`/`UnmarshalJSON`, and rely on carefully constructed functions (`ValueToInterface`, and friends) and the `ast.Term` type for serialization. In fact, I have to wrap the `ast.Object` in a `Term` in order for it to serialize correctly. :thinking:

Reply:
Hah, I see. Can we push `c` through `ValueToInterface`? 🤔 But this is ok, too, I think.